### PR TITLE
[ONNX] De-duplicate initializers (#68202)

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69550
* #69548
* __->__ #69547
* #72663
* #68491

ScriptModule export introduces duplicated ONNX initializers for shared weights, unnecessarily increases ONNX model size. This PR de-duplicates ONNX initializers for model exported in eval mode, by checking if the underlying tensors share the same `data_ptr`, `strides` and `sizes`.

Co-authored-by: BowenBao <bowbao@microsoft.com>

Differential Revision: [D32994271](https://our.internmc.facebook.com/intern/diff/D32994271)

### CODE REVIEW

**Comment by msaroufim** [comment]
@msaroufim has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32994271).

File: torch/csrc/jit/passes/onnx/helper.h [hunk 1]
```diff
@@ -14,6 +14,8 @@ static const int OPSET_VERSION_10 = 10;
 static const int OPSET_VERSION_11 = 11;
 static const int OPSET_VERSION_12 = 12;
 static const int OPSET_VERSION_13 = 13;
+static const int OPSET_VERSION_14 = 14;
+static const int OPSET_VERSION_15 = 15;
```
Comment (malfet) [inline torch/csrc/jit/passes/onnx/helper.h:18]: BE question: shouldn't it be enum?
```suggestion
enum {
  OPSET_VERSION_11 = 11,
  OPSET_VERSION_12,
  OPSET_VERSION_13,
  OPSET_VERSION_14,
  OPSET_VERSION_15,
};
```

File: torch/onnx/__init__.py [hunk 2]
```diff
@@ -10,7 +10,7 @@
 
 producer_name = "pytorch"
 producer_version = _C._onnx.PRODUCER_VERSION
-constant_folding_opset_versions = [9, 10, 11, 12, 13, 14]
+constant_folding_opset_versions = [9, 10, 11, 12, 13, 14, 15]
```
Comment (malfet) [inline torch/onnx/__init__.py:13]: Could this list ever be contiguous? If not, lets ask python to generate it
```suggestion
constant_folding_opset_versions = list(range(9, 16))
```

**Comment by msaroufim** [comment]
@msaroufim has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32994271).

**Comment by msaroufim** [comment]
@msaroufim has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32994271).

**Comment by msaroufim** [comment]
@msaroufim has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32994271).

**Comment by malfet** [comment]
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32994271).

**Pull request closed without merging**